### PR TITLE
Introduce version to comparison number

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
To do the work for [AB#8548](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8548) I've introduced this comparison algorithm.

This very simply and dumbly solves the problem of version comparison.

### CODE REVIEW

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 1]
```diff
@@ -589,3 +589,61 @@ export function versionHasMovedSparsedMatrix(version: string): boolean {
 		version >= "2.0.0-internal.2.0.0" || (!version.includes("internal") && version >= "2.0.0")
 	);
 }
+
+
+/**
+ * @internal
+ */
+export function versionToComparisonNumber(version: string): number {
```
Comment (tyler-cai-microsoft) [inline packages/test/test-version-utils/src/versionUtils.ts:597]: @tylerbutler , do we already have something like this?

**Review by jatgarg — APPROVED** [review]

File: packages/test/test-version-utils/src/versionUtils.ts (hunk 1 above)
Comment (tylerbutler) [reply packages/test/test-version-utils/src/versionUtils.ts:596]: What's the intended use of this? Could you add some docs?

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 2]
```diff
@@ -589,3 +589,60 @@ export function versionHasMovedSparsedMatrix(version: string): boolean {
...
+	if (version.startsWith("0.")) {
+		return 0;
+	}
+	if (version.startsWith("1.")) {
+		return 1;
+	}
+	if (version.startsWith("2.0.0-internal.1")) {
+		return 2;
+	}
```
Comment (tylerbutler) [inline packages/test/test-version-utils/src/versionUtils.ts:605]: You're basically building a map of starting strings --> indices. So you could just have an array of strings, and the value you return is the index of the first matching startsWith in the array.

```ts
const versionOrder["0.", "1." ...];
return versionOrder.findIndex((prefix) => version.startsWith(prefix))
```

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 3]
```diff
@@ -589,3 +589,60 @@ export function versionHasMovedSparsedMatrix(version: string): boolean {
...
+	if (version.startsWith("2.0.0-rc.5")) {
+		return 14;
+	}
+
+	const parsed = semver.parse(version);
+	if (!parsed) {
+		throw new Error(`Invalid version: ${version}`);
+	}
+	return parsed.major * 1_000_000 + parsed.minor * 1000 + parsed.patch + 15;
```
Comment (tylerbutler) [inline packages/test/test-version-utils/src/versionUtils.ts:647]: say what now?

Comment (tylerbutler) [reply packages/test/test-version-utils/src/versionUtils.ts:647]: Before adding more code to test-version-utils, I think we should discuss what we need that package to do and what version schemes we expect it to handle. We've tried the "fix it as we go" approach and it hasn't worked.

Ultimately I think we're looking for a way to compare and sort versions of different schemes, but that's just based on the problems that I have seen with compat tests and version bumps. IMO, not all of these combinations can be supported. But regardless, we need to define what version schemes we need to compare then write test cases and go from there.

**Pull request merged**